File: tracktime/entry_list.py
import csv
import os

from pathlib import Path
from typing import Any, Dict, Union

from tracktime.time_entry import TimeEntry
from datetime import timedelta, date, datetime
from tracktime.time_parser import parse_time


def get_path(
    config: Dict[str, Any],
    date: Union[date, datetime],
    makedirs: bool = False,
) -> Path:
    """
    Returns the path for a given date.

    Arguments:
    date: the date[time] object representing the date to get a path for.
    makedirs: (optional) whether or not to create the corresponding directory

    Returns: ``Path`` for the date

    >>> from datetime import date
    >>> from tracktime.config import get_config
    >>> get_path(get_config(), date(2018, 1, 1)).parts[-3:] == (
    ... '2018', '01', '01')
    True
    """
    directory = Path(config["directory"])
    directory = directory.joinpath(str(date.year))
    directory = directory.joinpath("{:02}".format(date.month))

    if makedirs:
        os.makedirs(directory, exist_ok=True)

    return directory.joinpath("{:02}".format(date.day))


class EntryList:
    """
    A list of ``TimeEntry``s.

    Note: If customer is specified, the EntryList should only be used for
    display and saving is disabled due to the fact that it would override
    entries from other customers.
    """

    def __init__(self, config, date, customer=None):
        self.date = date
        self.config = config
        self.customer = customer
        self.entries = []

        # Load entries from the file
        self.filepath = get_path(self.config, date, makedirs=True)
        if os.path.exists(self.filepath):
            with open(self.filepath, "r") as f:
                for row in csv.DictReader(f):
                    # Convert times to datetimes
                    for k in ("start", "stop"):
                        if row[k]:
                            row[k] = parse_time(row[k])

                    if not self.customer or row["customer"] == self.customer:
                        self.entries.append(TimeEntry(**row))

    def __len__(self):
        return len(self.entries)

    def __getitem__(self, key):
        return self.entries[key]

    @property
    def total(self):
        total_minutes = sum(e.duration(allow_unended=True) for e in self.entries)
        return (total_minutes // 60, total_minutes % 60)

    def add_entry(self, entry):
        if self.customer:
            return

        # Determine where the entry does in the list.
        index = len(self.entries) + 1  # default to the end
        for i, e in enumerate(self.entries):
            if e.stop and e.start <= entry.start < e.stop:
                # The entry is being started in the middle of this one.
                entry.stop = e.stop
                e.stop = entry.start
                index = i + 1
                break

            if entry.start < e.start:
                # The entry is being started before this.
                entry.stop = e.start
                index = i
                break

            # There is an unended time entry. Stop it, and start the new one.
            if e.start <= entry.start and not e.stop:
                e.stop = entry.start
                index = i + 1
                break

        self.entries.insert(index, entry)

    def save(self):
        if self.customer:
            return

        with open(self.filepath, "w", newline="") as f:
            fieldnames = [
                "start",
                "stop",
                "type",
                "project",
                "taskid",
                "customer",
                "description",
            ]
            writer = csv.DictWriter(f, fieldnames=fieldnames)

            writer.writeheader()
            for entry in self.entries:
                writer.writerow(dict(entry))

    def save_and_sync(self):
        if self.customer:
            return

        self.save()
        self.sync()

    def sync(self):
        if self.customer:
            return

        from tracktime.synchronisers import Synchroniser

        Synchroniser(self.config).sync(date(self.date.year, self.date.month, 1))

    def start(self, start, description, type, project, taskid, customer):
        if self.customer:
            return

        time_entry = TimeEntry(
            start,
            description,
            type=type,
            project=project,
            taskid=taskid,
            customer=customer,
        )
        self.add_entry(time_entry)
        self.save_and_sync()

    def stop(self, stop):
        if self.customer:
            return

        if len(self) == 0 or self.entries[-1].stop:
            raise Exception("No time entry to end.")

        self.entries[-1].stop = stop
        self.save_and_sync()

    def resume(self, start, entry):
        if self.customer:
            return

        if len(self) == 0:
            yesterday = self.date - timedelta(days=1)
            old_entry = EntryList(self.config, yesterday).entries[-1]
        else:
            old_entry = self.entries[entry]

        self.start(
            start,
            old_entry.description,
            old_entry.type,
            old_entry.project,
            old_entry.taskid,
            old_entry.customer,
        )

File: lib/entrylist.go
package lib

import (
	"encoding/csv"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type TimeEntryType string

func ParseTimeEntryType(typeStr string) TimeEntryType {
	switch strings.ToLower(typeStr) {
	case "gh":
		return "github"
	case "gl":
		return "gitlab"
	default:
		return TimeEntryType(typeStr)
	}
}

type Customer string
type Project string
type TaskID string

type TimeEntry struct {
	Index       int
	Start       *Time
	Stop        *Time
	Type        TimeEntryType
	Project     Project
	Customer    Customer
	TaskID      TaskID
	Description string
}

func NewEntryFromRecord(idx int, record []string) (te *TimeEntry, err error) {
	te = &TimeEntry{Index: idx}
	te.Start, err = ParseTime(record[0])
	if err != nil {
		return
	}
	te.Stop, err = ParseTime(record[1])
	if err != nil {
		return
	}
	te.Type = ParseTimeEntryType(record[2])
	te.Project = Project(record[3])
	te.TaskID = TaskID(record[4])
	te.Customer = Customer(record[5])
	te.Description = record[6]
	return
}

func (te *TimeEntry) Duration(allowUnended bool) (time.Duration, error) {
	stop := te.Stop
	if stop == nil {
		if allowUnended {
			stop = CurrentTime()
		} else {
			return time.Duration(0), fmt.Errorf("unended time entries cannot have a duration")
		}
	}
	return stop.Sub(te.Start), nil
}

type EntryList struct {
	Date    Date
	Config  *Config
	entries []*TimeEntry
}

func DayFilename(config *Config, date Date) string {
	return filepath.Join(
		config.Directory.Expand(),
		date.Format("2006"),
		date.Format("01"),
		date.Format("02"),
	)
}

func EntryListForDay(config *Config, date Date) (*EntryList, error) {
	file, err := os.OpenFile(DayFilename(config, date), os.O_RDONLY|os.O_CREATE, 0644)
	if err != nil {
		return nil, err
	}

	el := EntryList{Date: date, Config: config}
	reader := csv.NewReader(file)
	i := 0
	for {
		record, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}
		if i == 0 {
			i++
			continue
		}

		entry, err := NewEntryFromRecord(i, record)
		if err != nil {
			return nil, err
		}
		el.entries = append(el.entries, entry)
		i++
	}

	return &el, nil
}

func (el *EntryList) EntriesForCustomer(customer Customer) []*TimeEntry {
	if len(customer) == 0 {
		return el.entries
	}

	entries := []*TimeEntry{}
	for _, e := range el.entries {
		if e.Customer == customer {
			entries = append(entries, e)
		}
	}
	return entries
}

func (el *EntryList) TotalTimeForCustomer(customer Customer) time.Duration {
	minutes := time.Duration(0)
	for _, e := range el.entries {
		if len(customer) == 0 || e.Customer == customer {
			duration, _ := e.Duration(true)
			minutes += duration
		}
	}
	return minutes
}

func (el *EntryList) AddEntry(entry *TimeEntry) {
	insertIdx := len(el.entries) + 1
	for i, e := range el.entries {
		if e.Stop != nil && entry.Start.Between(e.Start, e.Stop) {
			// The entry is being started in the middle of this one
			entry.Stop = e.Stop
			e.Stop = entry.Start
			insertIdx = i + 1
			break
		}

		if entry.Start.Before(e.Start) {
			// The entry is being started before this.
			entry.Stop = e.Start
			insertIdx = i
			break
		}

		if e.Start.Before(entry.Start) && e.Stop == nil {
			// There is an unended time entry. Stop it, and start the new one.
			e.Stop = entry.Start
			insertIdx = i + 1
			break
		}
	}

	newEntries := make([]*TimeEntry, len(el.entries)+1)
	for i := 0; i < len(newEntries); i++ {
		if i < insertIdx {
			newEntries[i] = el.entries[i]
		} else if i == insertIdx {
			newEntries[i] = entry
		} else {
			newEntries[i] = el.entries[i-1]
		}
	}
	el.entries = newEntries
}

func (el *EntryList) Save() error {
	file, err := os.OpenFile(DayFilename(el.Config, el.Date), os.O_WRONLY|os.O_CREATE, 0644)
	if err != nil {
		return err
	}

	writer := csv.NewWriter(file)
	err = writer.Write([]string{"start", "stop", "type", "project", "taskid", "customer", "description"})
	if err != nil {
		return err
	}

	for _, e := range el.entries {
		err = writer.Write([]string{
			e.Start.String(),
			e.Stop.String(),
			string(e.Type),
			string(e.Project),
			string(e.TaskID),
			string(e.Customer),
			e.Description,
		})
		if err != nil {
			return err
		}
	}
	writer.Flush()
	return writer.Error()
}

func (el *EntryList) Sync() error {
	// TODO
	return nil
}

func (el *EntryList) SaveAndSync() error {
	err := el.Save()
	if err != nil {
		return err
	}
	return el.Sync()
}

func (el *EntryList) Start(start *Time, description string, taskType TimeEntryType, project Project, customer Customer, taskID TaskID) error {
	newEntry := &TimeEntry{
		Start:       start,
		Type:        taskType,
		Project:     project,
		Customer:    customer,
		TaskID:      taskID,
		Description: description,
	}
	el.AddEntry(newEntry)
	return el.SaveAndSync()
}

func (el *EntryList) Stop(stop *Time) error {
	if len(el.entries) == 0 || el.entries[len(el.entries)-1].Stop != nil {
		return fmt.Errorf("no time entry to stop")
	}
	el.entries[len(el.entries)-1].Stop = stop
	return el.SaveAndSync()
}

func (el *EntryList) Resume(resumeIndex int, description *string, start *Time) error {
	var oldEntry *TimeEntry
	if resumeIndex == -1 {
		if len(el.entries) > 0 {
			oldEntry = el.entries[len(el.entries)-1]
		} else {
			yesterdayEntries, err := EntryListForDay(el.Config, el.Date.AddDays(-1))
			if err != nil {
				return err
			}
			if len(yesterdayEntries.entries) == 0 {
				return fmt.Errorf("no time entry to resume")
			}
			oldEntry = yesterdayEntries.entries[len(yesterdayEntries.entries)-1]
		}
	} else {
		oldEntry = el.entries[resumeIndex-1]
	}

	if description == nil {
		description = &oldEntry.Description
	}
	return el.Start(start, *description, oldEntry.Type, oldEntry.Project, oldEntry.Customer, oldEntry.TaskID)
}
